### PR TITLE
Make repos2bow work on PGA

### PR DESCRIPTION
**Pull request by r0mainK**
### CONTEXT:

So, again on this. Thought the previous PR was enough, but there were two shortcoming:
- we still had to run repos2bow subdir by subdir, either manually or through bash loops
- on PGA still got OOM errors

### CHANGES:

**1st commit:**

I added broadcasting in `TFIDF` and `Indexer`transformers, to speed up process and avoid OOM. I removed `cached_index_path`  in `Indexer`'s `__init__` because it was kind of redundant with the `column2id` param. I also replaced in both transformers the use of the model with dicts, because the broadcasting of whole models is more costly.

**2nd commit:**

Even though we can't hold all extracted features at one or we get OOM errors, we can save on disk all of PGA and then iterate on partitions to run the pipeline - as long as TFIDF is done intelligently. To do so we partition the data by document with `PartitionBy` transformer I added, then select by partition number with `PartitionSelector`. 

**3rd commit:**

I use the new or modified transformers in repos2bow. Now `repos2bow_index` must have been run previously to run `repos2bow`, I added `ValueErrors` to `docfreq`and `quant` utilitaries. To choose the number of iterations I added a `num-iterations` flag. I added an `uncache`  method to `Cacher` so Spark doesn't waste memory on data cached in a previous iteration.

The pipeline is now:

1. Load `docfreq.asdf`, `index.asdf`, `quant.asdf` if needed, all created w/ `repos2bow_index`.
2. `PartitionBy` on all the data then cache it on disk (takes about an hour for PGA divided in 500)
3. Iterate through all partitions:

```
# first write meta table
INFO:MetadataSaver:ParquetLoader -> Moder -> PartitionBy -> Cacher -> PartitionSelector 
-> Repartitioner -> Cacher -> MetadataSaver

# then collect documents in that partition to reduce size of broadcasted doc index:
INFO:Collector:ParquetLoader -> Moder -> PartitionBy -> Cacher -> PartitionSelector 
-> Repartitioner -> Cacher -> UastRow2Document -> Cacher ->  FieldsSelector 
-> Distinct -> Collector

# then collect distinct tokens to reduce size of broadcasted token_index / token df
INFO:Collector:ParquetLoader -> Moder -> PartitionBy -> Cacher -> PartitionSelector 
-> Repartitioner -> Cacher -> UastRow2Document -> Cacher -> UastDeserializer 
-> Uast2BagFeatures -> BagFeatures2TermFreq -> Cacher -> FieldsSelector 
-> Distinct -> Collector

# then write bags
INFO: BOWWriter:ParquetLoader -> Moder -> PartitionBy -> Cacher -> PartitionSelector 
-> Repartitioner -> Cacher -> UastRow2Document -> Cacher -> UastDeserializer 
-> Uast2BagFeatures -> BagFeatures2TermFreq -> Cacher -> TFIDF -> Indexer (doc)
-> Indexer (token) -> Repartitioner -> BagsSaver -> PartitionBy -> BOWWriter
```

**RESULTS ON PGA:**

I am running with 500 iterations, it took about an hour to cache all the data to disk (197GB), then each iteration runs in about 10 minutes, and writes 170MB -> job should run in ~85h and write 85GB to disk

### CODE REVIEW

File: sourced/ml/cmd/repos2bow.py [hunk 1]
```diff
@@ -4,50 +4,99 @@
...
+            .link(document_indexer) \
+            .link(Indexer(Uast2BagFeatures.Columns.token, root.session.sparkContext,
+                          reduced_token_index))
+        if save_hook is not None:
+            bags_writer = bags_writer \
+                .link(Repartitioner.maybe(args.partitions, args.shuffle, multiplier=10)) \
+                .link(save_hook())
+        bow = args.bow.split(".asdf")[0] + "_" + str(num_part + 1) + ".asdf"
+        bags_writer \
+            .link(PartitionBy.maybe(
```
Comment (vmarkovtsev) [inline sourced/ml/cmd/repos2bow.py]: "PartitionBy" should be renamed to "Partitioner"

File: sourced/ml/models/df.py [hunk 2]
```diff
@@ -58,6 +58,13 @@ def docs(self) -> int:
         """
         return self._docs
 
+    @property
+    def df(self) -> dict:
```
Comment (vmarkovtsev) [inline sourced/ml/models/df.py]: Now I see. We need to put a huge comment which restricts from exposing the dictionary directly.

File: sourced/ml/cmd/args.py [hunk 3]
```diff
@@ -126,6 +126,9 @@ def add_bow_args(my_parser: argparse.ArgumentParser):
     my_parser.add_argument(
         "--batch", default=BOWWriter.DEFAULT_CHUNK_SIZE, type=int,
         help="The maximum size of a single BOW file in bytes.")
+    my_parser.add_argument(
+        "--num-iterations", default=1, type=int,
+        help="The number of iterations required to parse the data.")
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply sourced/ml/cmd/args.py]: Yeah the formulation is not great, basically after partioning by document I run the pipeline on each partition separately within a loop. This argument indicates the number of partitions.

File: sourced/ml/cmd/repos2bow.py (hunk 1 above)
Comment (r0mainK) [reply sourced/ml/cmd/repos2bow.py]: done

File: sourced/ml/cmd/repos2bow.py [hunk 4]
```diff
@@ -4,50 +4,99 @@
...
-        .link(document_indexer) \
-        .link(Indexer(Uast2BagFeatures.Columns.token, df_model.order))
-    if save_hook is not None:
-        bags_writer = bags_writer \
-            .link(Repartitioner.maybe(args.partitions, args.shuffle, multiplier=10)) \
-            .link(save_hook())
-    bags_writer.link(BOWWriter(document_indexer, df_model, args.bow, args.batch)) \
-        .execute()
+    log.info("Loading the document index from %s ...", args.cached_index_path)
+    document_index = DocumentFrequencies().load(source=args.cached_index_path).df
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply sourced/ml/cmd/repos2bow.py]: so from your other comment i don't know if you still want me to do this or if the comment is enough ?

File: sourced/ml/models/df.py (hunk 2 above)
Comment (r0mainK) [reply sourced/ml/models/df.py]: okay, adding

**Comment by vmarkovtsev** [comment]
Coverage to be improved in https://github.com/src-d/ml/issues/300

**Pull request merged**